### ISSUE TITLE
Interface and Implementation with different otptional param value should generate a warning.

### ISSUE THREAD

**Issue by marcelodeaguiar** [dotnet/roslyn#8104]
Consider the following code:

```
public interface IFoo
{
    bool Shine(string name, bool optionalBool = false);
}

public class Mirror : IFoo
{
    public bool Shine(string name, bool optionalBool = true)
    {
        return optionalBool;
    }
}

public class Test
{
    public Test()
    {
        IFoo foo = new Mirror();
        bool theBoolValue = foo.Shine("name");

        System.Diagnostics.Debug.WriteLine(theBoolValue);
    }
}
```

One should expect the value to be printed in the debug console to be **true** but the value printed is **false**. That is because the generated code by the compiler uses the interface optional value instead on the class even thou we are instantiating the class explicetly. My guess is that this happens because I'm using the interface to call the method, is that assumption true? Is this the expected bahavior?

The fact the value is different in the interface and the implementation does not mean the contract is broken?

Anyway, the compiler should generate at least a warning that the value in the interface is different from the implementation as this is a really hard problem to find and really common if you use optional on interfaces.

**Comment by HaloFour** [comment]
Optional values aren't a part of the method signature.  They are additional metadata tacked onto the parameter.  That metadata is ignored by the CLR at runtime.  It is only ever used by a compiler that supports it in order to silently embed that optional value as the argument.  So it makes sense that the value you are seeing embedded is the value associated with the method in the interface, not the value associated with the overridden method in the concrete class, because the compiler is resolving the value based on the type of the variable.  If you were to change `foo` to a the type `Mirror` then you'd see that the `optionalBool` parameter is being set to `true`.

**Comment by gafter** [comment]
@marcelodeaguiar We invite you to write an analyzer to help diagnose this.

**Comment by srivatsn** [comment]
Closing this issue per @gafter's suggestion.
